include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+
+  nonlinear_solver.solve(initial_guess_and_solution);
+}
+#endif
+
+
+
+template <typename VectorType>
+void
+NonlinearSolverSelector<VectorType>::solve(
+  VectorType &initial_guess_and_solution)
+{
+  // The "auto" solver_type will default to kinsol, however if KINSOL is not
+  // available then we will use NOX.
+  if (additional_data.solver_type == AdditionalData::SolverType::automatic)
+    {
+#ifdef DEAL_II_WITH_PETSC
+      additional_data.solver_type = AdditionalData::SolverType::petsc_snes;
+#endif
+#ifdef DEAL_II_WITH_TRILINOS
+      additional_data.solver_type = AdditionalData::SolverType::nox;
+#endif
+#ifdef DEAL_II_WITH_SUNDIALS
+      additional_data.solver_type = AdditionalData::SolverType::kinsol;
+#endif
+
+      // If "auto" is still the solver type we cannot solve the problem
+      if (additional_data.solver_type == AdditionalData::SolverType::automatic)
+        AssertThrow(false, ExcMessage("No valid solver type."));
+    }
+
+  if (additional_data.solver_type == AdditionalData::SolverType::kinsol)
+    {
+#ifdef DEAL_II_WITH_SUNDIALS
+      SUNDIALS::KINSOL<VectorType> nonlinear_solver(additional_data_kinsol,
+                                                    mpi_communicator);
+
+      // We set the KINSOL reinit vector equal to the same function
+      // defined for NonlinearSolverSelector.
+      nonlinear_solver.reinit_vector = reinit_vector;
+
+      nonlinear_solver.residual = residual;
+
+      // We cannot simply set these two functions equal to each other
+      // because they have a different number of inputs.
+      nonlinear_solver.setup_jacobian = [&](const VectorType &current_u,
+                                            const VectorType /*&current_f*/) {
+        return NonlinearSolverSelector<VectorType>::setup_jacobian(current_u);
+      };
+
+      nonlinear_solver.solve_with_jacobian = solve_with_jacobian;
+
+      nonlinear_solver.solve(initial_guess_and_solution);
+#else
+      AssertThrow(
+        false, ExcMessage("You do not have SUNDIALS configured with deal.II!"));
+#endif
+    }
+  else if (additional_data.solver_type == AdditionalData::SolverType::nox)
+    {
+#ifdef DEAL_II_WITH_TRILINOS
+      TrilinosWrappers::NOXSolver<VectorType> nonlinear_solver(
+        additional_data_nox, parameters_nox);
+
+      // Do the same thing for NOX that we did with KINSOL.
+      nonlinear_solver.residual = residual;
+
+      // setup_jacobian for NOX has the same number of arguments for the same
+      // function in NonlinearSolverSelector.
+      nonlinear_solver.setup_jacobian = setup_jacobian;
+
+      nonlinear_solver.solve_with_jacobian = solve_with_jacobian;
+
+      nonlinear_solver.solve(initial_guess_and_solution);
+#else
+      AssertThrow(
+        false, ExcMessage("You do not have Trilinos configured with deal.II"));
+#endif
+    }
+  else if (additional_data.solver_type ==
+           AdditionalData::SolverType::petsc_snes)
+    {
+      solve_with_petsc(initial_guess_and_solution);
```

Comment:
Out of curiosity, what's the reason why you broke this into a separate function, but not the other places? Because you can only support certain vector types? Perhaps put that into a short comment?